Refactors reconcile tasks to split out tree parsing

tbranyen opened the pull request:
I've added a diffhtml-bundling repo in this PR to demonstrate bundling/building and running the various flavors of diffHTML. Its just a start and I may remove it from the end and turn into documentation. For now its serving as a good reference to test things.

Instead of writing two different versions of the reconcile-trees task, I've broken out the parsing into a separate task. I've set up default (runtime) tasks in the transaction object. The main build augments this array to inject in parsing.

mAAdhaTTah commented on packages/diffhtml-components/lib/component.js:
```diff
@@ -1,7 +1,7 @@
+import process from './util/process';
```
Can't we import `./util/process` as `import { process } from 'diffhtml-shared-internals';`? There are a couple imports throughout that can be rewritten this way, which may help solve some of the issues you're having.

tbranyen replied:
`./util/process` has `import { process } from 'diffhtml-shared-internals/lib/process';`

The util simply makes sure that `process` is a global.

mAAdhaTTah commented on packages/diffhtml-middleware-inline-transitions/index.js:
```diff
@@ -1,8 +1,4 @@
-  removeTransitionState,
-} from 'diffhtml/lib/transition';
-
+const { addTransitionState, removeTransitionState } = require('diffhtml');
```
Why switch this to CJS?

tbranyen replied:
Whoops, good catch

mAAdhaTTah commented on packages/diffhtml-middleware-verify-state/.babelrc:
```diff
@@ -1,6 +1,23 @@
+        ["module-rewrite", { "replaceFunc": "../../utils/replace-es.js" }]
+      ]
+    }
+  }
```
If all of these babel configurations are the same, is there a way to share them?

tbranyen replied:
Not sure, do you have any suggestions? Our React components repo at work uses a custom tool to abstract away shared config, but that would be too complex to bring into this repo I think.

tbranyen replied:
I'm working on a babel preset to minimize the amount of configuration, so far so good :)

mAAdhaTTah replied:
Yeah, I was thinking a preset, although I'm not sure how that would work with the `replaceFunc` configuration.

mAAdhaTTah commented on packages/diffhtml/lib/runtime.js:
```diff
@@ -13,9 +7,6 @@ import use from './use';
-  schedule, shouldUpdate, reconcileTrees, syncTrees, patchNode, endAsPromise,
-];
 
 const innerHTML = bindInnerHTML(defaultTasks);
```
`defaultTasks` isn't available here.

tbranyen replied:
Yup, gotta fix that.

mAAdhaTTah commented on packages/diffhtml/lib/use.js:
```diff
@@ -27,7 +27,7 @@ export default function use(middleware) {
   // Call the subscribe method if it was defined, passing in the full public
   // API we have access to at this point.
-  subscribe && middleware.subscribe(use.diff);
+  subscribe && middleware.subscribe();
```
So middleware is no longer passed the internals, those have to be imported?

tbranyen commented:
To answer your question regarding browserify or rollup, I'd prefer to use rollup everywhere. It's a bit more complex, but it does the job better.

tbranyen commented:
This is moving in the right direction for separating internals @mAAdhaTTah. I'm just trying to trace the module path for each. I think I may end up moving towards a shared symbol. Some thoughts on how that might work:

diffhtml-shared-internals in CJS/ES environments will return an object that
has all the internals assigned to it.

in UMD (browser) environments, this module will attach the same object to the
global scope using a symbol, and will be baked into diffHTML core (to avoid
a separate script or version confusion). the shared module will export this
symbol and the middleware code will be rewritten in UMD land (using Rollup) to
import the symbol and reference the global

The following shows the input and output 

ES (input, output contains a shared reference baked into the bundle):

``` js
import { process } from 'diffhtml-shared-internals/lib/process';
```

CJS (input, output contains a shared reference baked into the bundle):

``` js
const { process } = require('diffhtml-shared-internals/lib/process');
```

UMD (output, the input is going to be source ESM):

``` js
var process = global[__difhtml__shared__internals.$$symbol].process;
```

I'm also considering refactoring out transitions completely into middleware, making diffHTML 100% synchronous and no Promises baked in. I think this will shave off quite a bit of filesize.

mAAdhaTTah commented:
If the purpose of `shared-internals` is to provide a consistent interface for importing the internals, then we should really avoid this:

```js
import { process } from 'diffhtml-shared-internals/lib/process';
```

in favor of this:

```js
import { process } from 'diffhtml-shared-internals';
```

Otherwise, you're no better off than just doing this:

```js
import { process } from 'diffhtml/lib/util/process';
```

mAAdhaTTah commented:
I'm also really confused as to why this whole global symbol thing is necessary. It seems like an overcomplication compared to being table to solve this in the UMD bundler. What issue are you running into that this solves?

tbranyen commented:
@mAAdhaTTah Going to separate my feedback to your comments from the intended goals:

- I want a consumer to be able to pick from either UMD, CJS, or ESM and have all official middleware/plugins work in harmony with that decision.
- I want consumers to be able to run diffHTML in either a main (full/default) or a runtime (barebones) build.
- I want the consumer to be able to pick arbitrary middleware that "just works" with the current build they selected.
- If a developer chooses to access internals, it's through a "safe" abstraction that will ensure they will hardly ever need to rewrite their module. This is very similar in philosophy to the new experimental Node addon api: https://github.com/nodejs/node-addon-api

---

I can't think of another solution, and your proposal to use `diff.Internals` doesn't make sense given how diffHTML works. It's not like React in that it has a single global namespace. Due to how JSX works, users are required to import the `React` namespace in its entirety. diffHTML doesn't require this, which is why you can import only the functions you need.

If you wanted to access the internals of diffHTML, you cannot simply just import `diffhtml` since you don't know if the user is running the runtime build or the main build. It could be possible to use a webpack/browserify alias to solve this, but that is significantly more work for the consumer. I'd rather push the complexity to diffHTML.

To answer your question, why are we importing the `lib` code vs the main module? Either will work! Choose which one is best for you, but understand that tree-shaking may not necessarily work the way you'd expect by importing like that. I choose to import from `lib` specifically, in order to get the exact module I'm looking for and nothing more from any optimizer/bundler. This even works to effectively "tree-shake" in browserify under CJS. It's absolutely nothing like importing the internals directly from diffHTML. If you do that and I start changing stuff around there's no guarantee your code will still work. I will attempt to guarantee that `diffhtml-shared-internals` will not change its API surface area between say 1.0 and 2.0.

To be honest, the global symbol seems like a perfect solution and since it's 100% hidden from the user, it makes good on the `diffhtml-shared-internals` promise of being a suitable abstraction. Since you won't need to change your code, and this approach still works, I think it's fine.

I'm going to be adding a bunch of examples of using diffHTML under various configurations in the diffhtml-bundling package. Right now it just has web globals, and if you can think of a smarter/better to achieve the same goals with less complexity, I'm :100:% for it!

mAAdhaTTah commented:
This is great, now that I have a sense of what you're going for. I'm primarily coming at this from the perspective of a user consuming this module through a bundler; it sounds like you're pretty keen on making sure this works for UMD builds as well, so I think we should be able to find something that works for both cases.

> To answer your question, why are we importing the lib code vs the main module? Either will work! Choose which one is best for you, but understand that tree-shaking may not necessarily work the way you'd expect by importing like that.

I guess the thing I'm worried about is a consumer using webpack/browserify and possibly ending up with two different instances of the internals in their final bundle. The consumer can't pull in specific paths from `shared-internals` the same way as the modules themselves do. The modules rewrite their import process; are we expecting people to use our preset as well? If we treat `lib` as a "public API" of sorts, we risk consumers ending up with 2 instances of the `shared-internals` and weird, hard-to-debug errors resulting from getting different instances of the `cache` module (which happened to me in an earlier iteration).

I definitely like having diffHTML absorb this complexity. I think the best way to ensure that doesn't happen is to have all the imports pointing to the package directly though. This has the added benefit of giving the consumer complete control over how everything gets resolved as well, so they can import any package from the direct export and pick and choose which bundle and apply whatever transforms are needed.

Overall, I guess I'm just a little concerned there's a lot of additional complexity that could make dealing with this with a bundler more difficult. Admittedly, I don't have any proof of that, so it's entirely possible the concern is unfounded, but the multi-internals problem.

I think I understand the `Symbol` solution; that seems fine, although it may be possible for a user to get nothing if the `shared-internals` are imported/accessed before `diffhtml` has added the internals to `global`.

The only other solution off the top I would suggest is actually putting the code from these internals into the `shared-internals` module directly, which then is responsible for adding whatever to `global`. Then its less of a wrapper or more actually where the code lives. It definitely makes dealing with it from a bundler perspective easier; not sure if that makes the UMD side harder though.

If you get a couple examples set up, we could probably play with all the options and see which of them work, even trying it with a few different bundlers (webpack, browserify, & rollup) to make sure they work without any specific configuration.

tbranyen commented:
Major issue with extracting the code out of diffHTML and into the shared repo is the added complexity with UMD builds. You now need to include a `diffhtml-internals` as well as `diffhtml`, and need to be very careful that multiple versions of that dependency aren't built. With the current Symbol approach, if multiple copies of shared-internals are loaded, its fine, they all point to the same object internally.

You are right about needing diffHTML loaded first before internals are accessed with this approach. Currently, I think this is a fair compromise and will document it like that for now. This will also mean its not possible to simply include internals in a Web Worker without a full fledged instance of diffHTML running there as well.

I think it'd be possible to do lazy-loading of functions to wait for diffHTML to be loaded. Like wrapping `createNode` with:

``` js
function createNode() {
  if (!internals.createNode) { throw 'error if not in production'; }

  return internals.createNode();
}
```

That would at least allow you to reference any internal function regardless if it was exported or not yet. You wouldn't be able to call it until diffHTML was loaded though.

mAAdhaTTah commented:
If we need to load diffHTML in order to make `shared-internals` work, my previous idea doesn't seem so bad:

> I can't think of another solution, and your proposal to use `diff.Internals` doesn't make sense given how diffHTML works. It's not like React in that it has a single global namespace.

FWIW, the idea was to be able to do this:

```js
import { Internals } from 'diffhtml';

// not this

import diff from 'diffhtml';

diff.Internals;
```

You don't have to import the whole namespace. `Internals` is a "subnamespace"; since you need to pull in the entirety of diffHTML to make the `shared-internals`work anyway, I don't think you're benefitting at all (from a tree-shaking perspective), and you end up with a single location to reference the cache. You wouldn't / shouldn't need to import subfolders either, eliminating the possibility of getting multiple instances.

The other possibility is to just export each of them through the index, e.g.:

```js
import { createNode, Pool, ...orAnyOfTheOthers } from 'diffhtml';
```

It would expand the API surface area (although to some extent, we're already doing that), but it could simplify a lot of this, rather than having to deal with this separate package.

tbranyen commented:
I think you're right about this. I wanted it to work such that consumers would never have to configure aliases to use the runtime build, but in the end it would be more simple and consistent. I really think we need to proxy the `Internals` though through `diffhtml-shared-internals`.

So basically it'd work like you describe, users could access `import { Internals } from 'diffhtml';` but the recommended way would be to use the abstraction. The abstraction would simply import the `Internals` and expose them in a safe way.

I don't want to clutter diffHTML core with code to support Internals in the future when we want to make breaking changes.

tbranyen commented:
I'll need to think about how a user will write:

``` js
import { caches } from 'diffhtml-shared-internals';
```

and have this compiled (in UMD) to:

``` js
var caches = diff.Internals.caches;
```

So basically the easiest way I can think of would be to use the hypotethical plugin for rollup, but that would mean all plugin authors would have to do the same thing, not really ideal. Maybe we present both options to a plugin author and explain the differences.

Unsafe way (will need to update code if Internals changes): Just import `Internals` from `diffhtml` directly. You are responsible for updating if the `Internals` API changes.

Safe way (no need to update code if Internals changes): Import your internals from `diffhtml-shared-internals` and use rollup w/ our plugin that rewrites the module definition for UMD builds. If they don't care about UMD then its not an issue.

mAAdhaTTah commented:
I'm not sure there's a benefit to even having that as a separate module though. In your "Safe vs Unsafe" examples, you can't (as the author of diffhtml) break the external API of the internal functions in either example. And in both examples, you should still be able to move files around without issue, as long as they're exported properly through the entry point. What protection does the module provide/enforce that we wouldn't be able to provide by treating the `Internals` the same way?

tbranyen commented:
Assuming we have `diff.Internals.parse` in 1.0 and we decide in 2.0 to remove parsing completely, I can't see how this wouldn't cause a rewrite/refactor for a end user's middleware. If they depend on shared-internals, we could just add the parsing there and their module maintains compatibility without core suffering the backwards-compat bloat.

If we expose `Internals` as a public core property you cannot simply dismiss the permanence of the API. For instance the core API of diffHTML has not really changed since I first wrote it. I've added a few methods, removed one, changed a few, but I'm trying to hold myself to not break end user code between updates. We need to scrutinize every single change that could affect the API otherwise you end up like React where some libs are locked to 0.14, because the internals changed so much in a breaking way that it's non-trivial to upgrade to 15. You can see react-blessed as a good example of this.

mAAdhaTTah commented:
> Assuming we have diff.Internals.parse in 1.0 and we decide in 2.0 to remove parsing completely, I can't see how this wouldn't cause a rewrite/refactor for a end user's middleware.

That's seems like an acceptable breaking change, since no dependency will be automatically updated, and we'd be able to move the parsing out to a separate library the user can pull in. It wouldn't require a major rewrite; just moving around an import.

> If we expose Internals as a public core property you cannot simply dismiss the permanence of the API.

You still have to do that with a separate module; the only difference is you can remove/deprecate a piece of the API with a module without affecting userland, which you can't do exporting them directly. You still have to ensure the function signatures and semantics remain consistent in both, and I think the added complexity of this other module, the global Symbol, and the possible path mixture isn't really worth it.

Comparably, N-API would be an abstraction _over_ the the current internals, rather than being the internals. The `shared-internals` module isn't as fleshed out as the layer provided by N-API, because the diffhtml internals aren't as complicated as V8, so there's no need for an abstraction layer.

> We need to scrutinize every single change that could affect the API otherwise you end up like React where some libs are locked to 0.14, because the internals changed so much in a breaking way that it's non-trivial to upgrade to 15. You can see react-blessed as a good example of this.

I don't think this actually protects you from React-like fragmentation, because it doesn't provide an abstraction over the things that cause breaking changes. It only really provides an abstraction over the internal folder structure.

tbranyen commented:
I hear your concerns and I agree with your perspective. But I don't want to expose `Internals` directly as an API because it effectively destroys the point of the public API being small and manageable. Exporting it is endorsing it. Exposing it as a non-enumerable property with a Symbol key seems like a better approach, which `diffhtml-shared-internals` then references (by pulling in diffHTML completely). If you want access to it in your code without dealing with the shared-internals you can opt into it with:

``` html
const diff = require('diffhtml');
const Internals = diff[Symbol.for('diffhtml.internals')];
```

Would that be an acceptable solution? Things may break if you access directly between major versions, and I'll provide migration paths in `diffhtml-shared-internals` if you decide to opt into that.

tbranyen commented:
I'd be okay with changing the overall strategy of using the runtime as well by documenting the alias configuration necessary to go from `diffhtml` -> `diffhtml/runtime` globally in your app.
